src/protocol/codecs/topic.rs

```diff
@@ -0,0 +1,121 @@
+use std::io::{self, ErrorKind};
+
+use bytes::{Buf, BytesMut};
+use tokio_util::codec::Decoder;
+
+use crate::protocol::{
+    codecs::{
+        payload::{MsgOfInterest, Payload},
+        tagmsg::Tag,
+    },
+    invalid_data,
+};
+
+/// Topic is a key-value string pair.
+pub struct Topic {
+    /// Key.
+    pub key: String,
+
+    /// Value.
+    pub value: String,
+}
+
+#[derive(Default, Clone)]
+pub struct TopicCodec {
+    /// Represents a message payload type identifier.
+    pub tag: Option<Tag>,
+}
+
+impl TopicCodec {
+    fn parse_topics(&mut self, src: &mut BytesMut) -> Result<Vec<Topic>, io::Error> {
+        let num_topics = src.get_u8();
+        let mut topics = Vec::with_capacity(num_topics as usize);
+
+        for _ in 0..num_topics {
+            let key_len = src.get_u8() as usize;
+            if key_len > src.len() {
+                return Err(invalid_data!("invalid topic length"));
+            }
+            let key = src.copy_to_bytes(key_len);
```

Comment:
why not directly `to_vec()`?

Reply:
ah yeah, because of the length; in any case, we could follow up with `to_vec()` here too, no?

Reply:
Ah, right
Done